### PR TITLE
CCD concurrency issue

### PR DESCRIPTION
**Pull request by AlokDatta**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPET-384

### Change description ###
Changes are made to following services -

- Created new ccd core data service client. Once this is set existing other core data services will be removed
- RequestUpdateCallbackService has been modified to cater the changes
- AllTabServiceImpl has been modified to cater the changes

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java [hunk 1]
```diff
@@ -52,40 +47,54 @@ public class CourtNavCaseService {
...
+    private final CoreCaseDataService coreCaseDataService;
+    private final SystemUserService systemUserService;
 
     public CaseDetails createCourtNavCase(String authToken, CaseData caseData) throws Exception {
+
+        String authorisation = systemUserService.getSysUserToken();
+        String systemUpdateUserId = systemUserService.getUserId(authorisation);
+
         Map<String, Object> caseDataMap = caseData.toMap(CcdObjectMapper.getObjectMapper());
         log.info("****************Creating courtnav case***************");
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java:59]: where is this core data now?

File: src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java [hunk 2]
```diff
@@ -52,40 +47,54 @@ public class CourtNavCaseService {
...
+                true
             );
 
-        CaseDataContent caseDataContent = CaseDataContent.builder()
-            .eventToken(startEventResponse.getToken())
-            .event(Event.builder()
-                       .id(startEventResponse.getEventId())
-                       .build())
-            .data(caseDataMap).build();
+        CaseData caseDataUpdate = CaseUtils.getCaseDataFromStartUpdateEventResponse(startEventResponse, objectMapper);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java:74]: blank case data as you dont have case number

File: src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java [hunk 3]
```diff
@@ -116,33 +125,28 @@ && checkTypeOfDocument(typeOfDocument)) {
...
+            );
+
+            coreCaseDataService.submitUpdate(
+                authorisationToken,
+                eventRequestData,
+                caseDataContent,
                 caseId,
-                true,
-                caseDataContent
+                true
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java:149]: I dont understand this code change. Its vague coding.

File: src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java [hunk 4]
```diff
@@ -86,14 +91,27 @@ public CaseDetails createCourtNavCase(String authToken, CaseData caseData) throw
 
     public void uploadDocument(String authorisation, MultipartFile document, String typeOfDocument, String caseId) {
 
+        String authorisationToken = systemUserService.getSysUserToken();
+        String systemUpdateUserId = systemUserService.getUserId(authorisationToken);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/courtnav/CourtNavCaseService.java:95]: We need to use the authorisation from the method header

File: src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CitizenCoreCaseDataService.java [hunk 5]
```diff
@@ -110,92 +120,35 @@ private EventRequestData eventRequest(CaseEvent caseEvent, String userId) {
...
+
+            CaseData caseDataUpdated = CaseUtils.getCaseDataFromStartUpdateEventResponse(startEventResponse, objectMapper);
+            CaseDataContent caseDataContent = coreCaseDataService.createCaseDataContent(
+                startEventResponse,
+                caseDataUpdated
             );
-            Map<String, Object> caseDataMap = caseData.toMap(objectMapper);
-            Iterables.removeIf(caseDataMap.values(), Objects::isNull);
-            CaseDataContent caseDataContent = caseDataContent(startEventResponse, caseDataMap);
-            return submitUpdate(
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CitizenCoreCaseDataService.java:193]: @remya-james , what happened to removed codes

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseDocumentController.java [hunk 6]
```diff
@@ -105,6 +106,12 @@ public class CaseDocumentController {
     @Value("${citizen.url}")
     private String dashboardUrl;
 
+    @Autowired
+    CoreCaseDataService coreCaseDataService;
+
+    @Autowired
+    private SystemUserService systemUserService;
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseDocumentController.java:113]: Use of this import?

File: src/main/java/uk/gov/hmcts/reform/prl/handlers/CaseEventHandler.java [hunk 7]
```diff
@@ -5,13 +5,13 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.event.EventListener;
 import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.prl.clients.ccd.CoreCaseDataService;
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/handlers/CaseEventHandler.java:8]: You just copied the existing event from one file to another. It is not serving either the purpose or reflecting the changes intended.

File: src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CitizenCoreCaseDataService.java (hunk 5 above)
Comment (remya-james) [reply src/main/java/uk/gov/hmcts/reform/prl/services/citizen/CitizenCoreCaseDataService.java:193]: @AlokDatta This is replaced with the case data retrieved from CaseUtils.getCaseDataFromStartUpdateEventResponse(startEventResponse, objectMapper);
Lines 76-79

File: src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseDocumentController.java (hunk 6 above)
Comment (remya-james) [reply src/main/java/uk/gov/hmcts/reform/prl/controllers/citizen/CaseDocumentController.java:113]: @AlokDatta I have removed this.

File: src/main/java/uk/gov/hmcts/reform/prl/handlers/CaseEventHandler.java (hunk 7 above)
Comment (remya-james) [reply src/main/java/uk/gov/hmcts/reform/prl/handlers/CaseEventHandler.java:8]: @AlokDatta  The triggerEvent function in CoreCaseData Service is being passed a map containing additional values like taskList and respondentTaskList when called from the CaseEventHandler. These are not available in CaseData. Hence I wasn't sure if we could replace this with the CaseData from StartEventResponse.

File: src/main/java/uk/gov/hmcts/reform/prl/services/tab/alltabs/AllTabServiceImpl.java [hunk 8]
```diff
@@ -70,19 +78,52 @@ public void updateAllTabsIncludingConfTab(CaseData caseData) {
...
+            combinedFieldsMap.put(DATE_SUBMITTED_FIELD, caseData.getDateSubmitted());
+        }
+        if (caseData.getCourtName() != null) {
+            combinedFieldsMap.put(COURT_NAME_FIELD, caseData.getCourtName());
+        }
+        if (caseData.getCourtId() != null) {
+            combinedFieldsMap.put(COURT_ID_FIELD, caseData.getCourtId());
+        }
+        getDocumentsMap(caseData, combinedFieldsMap);
+        refreshCcdUsingEvent(caseData, combinedFieldsMap);
```
Comment (AlokDatta) [inline src/main/java/uk/gov/hmcts/reform/prl/services/tab/alltabs/AllTabServiceImpl.java:113]: We are submitting the event just one line below. how this line 113 is significant?

**Comment by AlokDatta** [comment]
Not required anymore. it has been taken care as part of different PR.

**Comment by AlokDatta** [comment]
#1129

**Pull request closed without merging**